Document.md

```diff
@@ -0,0 +1,174 @@
+
+Because the system call handler is dynamically linked to the customized system call handler during execution, we should backup original glibc system call handler by `dlsym()`.
+```c
+__attribute__((constructor)) static void setup(void)
+{
+    pgsize = getpagesize();
+    in_segment = 0;
+    batch_num = 0;
+
+    /* store glibc function */
+    real_writev = real_writev ? real_writev : dlsym(RTLD_NEXT, "writev");
+    real_shutdown =
+        real_shutdown ? real_shutdown : dlsym(RTLD_NEXT, "shutdown");
+    real_sendfile =
+        real_sendfile ? real_sendfile : dlsym(RTLD_NEXT, "sendfile");
+    real_send =
+        real_send ? real_send : dlsym(RTLD_NEXT, "send");
+
+    global_i = global_j = 0;
+}
+```
+
+## The event-driven framework in lighttpd
+### Observe system call by strace
+1. `epoll_wait`: wait I/O event ready.
+2. `accept`: accept and create a new socket fd for connection.
+3. `read`: read the request in socket, but request is not ready.
+4. `epoll_ctl`: because request is not ready, add socket fd to epoll.
+5. `read`: request is ready, read the request in socket.
+6. `lseek`: the request is requesting a web page, so `lseek` would set offset to the location where the web page file is stored.
+7. `read`: read the web page file.
+8. `writev`: write the web page file to socket and send to client.
+```
+...
+4005.609 (577.243 ms): lighttpd/14597 epoll_wait(epfd: 4<anon_inode:[eventpoll]>, events: 0x55e5030116d0, maxevents: 1025, timeout: 1000) = 1
+...
+4583.125 ( 0.002 ms): lighttpd/14597 accept4(fd: 3<socket:[631034]>, upeer_sockaddr: 0x7ffc3059b420, upeer_addrlen: 0x7ffc3059b3c4, flags: 526336) = 6
+4583.129 ( 0.001 ms): lighttpd/14597 read(fd: 6<socket:[631037]>, buf: 0x55e503048420, count: 8191)        = -1 EAGAIN (Resource temporarily unavailable)
+4583.131 ( 0.001 ms): lighttpd/14597 epoll_ctl(epfd: 4<anon_inode:[eventpoll]>, op: ADD, fd: 6<socket:[631037]>, event: 0x7ffc3059b47c) = 0
+...
+10155.837 ( 0.001 ms): lighttpd/14303 read(fd: 6<socket:[636120]>, buf: 0x55a045451630, count: 8191)        = 106
+10155.846 ( 0.001 ms): lighttpd/14303 lseek(fd: 15</home/qwe661234/ESCA/web/index.html>, whence: SET)       = 0
+10155.848 ( 0.001 ms): lighttpd/14303 read(fd: 15</home/qwe661234/ESCA/web/index.html>, buf: 0x55a04544e4f9, count: 464) = 464
+10155.850 ( 0.005 ms): lighttpd/14303 writev(fd: 6, vec: 0x7ffc27ee0110, vlen: 1)                           = 681
+...
+```
+### Command
+* Server
+> $ sudo perf trace ./downloads/lighttpd1.4-lighttpd-1.4.58/src/lighttpd -D -f downloads/lighttpd1.4-lighttpd-1.4.58/src/lighttpd.conf
+* Client
+> $ ab -n 10 -c 10 -k http://127.0.0.1:3000/
+
+## Reference 
+* [GitHub: ESCA](https://github.com/eecheng87/ESCA)
+* [ESCA thesis](https://eecheng87.github.io/ESCA/main.pdf)
+* [System Call Note](https://hackmd.io/@combo-tw/Linux-%E8%AE%80%E6%9B%B8%E6%9C%83/%2F%40a29654068%2FHyD4Lu_Dr)
```

Comment:
Don't reference to HackMD notes because they might be changed frequently and unexpectedly.